password-hash/src/lib.rs

```diff
@@ -41,35 +41,41 @@ extern crate alloc;
 #[cfg(feature = "std")]
 extern crate std;
 
-use core::{
-    convert::{TryFrom, TryInto},
-    fmt,
-};
+pub mod errors;
+
+mod ident;
+mod output;
+mod params;
+mod salt;
+mod value;
 
 pub use crate::{
     errors::{HashError, HasherError, VerifyError},
     ident::Ident,
     output::Output,
-    params::Params,
+    params::ParamsBuf,
     salt::Salt,
     value::{Decimal, Value, ValueStr},
 };
 
 pub use b64ct as b64;
 
-pub mod errors;
-
-mod ident;
-mod output;
-mod params;
-mod salt;
-mod value;
+use core::{
+    convert::{TryFrom, TryInto},
+    fmt::{self, Debug},
+};
 
 /// Separator character used in password hashes (e.g. `$6$...`).
 const PASSWORD_HASH_SEPARATOR: char = '$';
 
 /// Trait for password hashing functions.
 pub trait PasswordHasher {
+    /// Algorithm-specific parameters.
+    type Params: Clone
+        + Debug
+        + for<'a> TryFrom<&'a ParamsBuf<'a>, Error = HasherError>
+        + for<'a> TryInto<ParamsBuf<'a>, Error = HasherError>;
```

Comment:
I played with a few different combinations of these bounds and this was the best I could come up with.

I think the one other thing to potentially consider is removing the `Error` bound on the `TryInto` impl, which would allow for an `Into` impl to satisfy this bound.